### PR TITLE
Adding trace logging to Environment Manager

### PR DESCRIPTION
**Pull request by dhoehna**
This PR adds simple trace logging to Environment Manager.  Every API surface call has a simple trace logging method.

This PR also has some warning fixes.

### CODE REVIEW

File: dev/EnvironmentManager/API/Microsoft.Windows.System.EnvironmentManager.Insights.h
Comment (DrusTheAxe) [inline dev/EnvironmentManager/API/Microsoft.Windows.System.EnvironmentManager.Insights.h:39]: No need for scope to be std::wstring as all you're doing is malloc+memcpy(+free) for a string constant

SUGGESTION: Refactor as a conversation function

```
constexpr PCWSTR ScopeToString(winrt::...::Scope scope)
{
    if (scope == ...::Process)
    {
        return L"Process";
    }
    ...
}
```

and change this function to

```
static void LogWithScopeAndMessage(...)
{
    auto scopeAsString{ ScopeToString(scope) };
    TraceLoggingClassWrite("CustomEvent", TraceLoggingValue(scopeAsString, "scope") ...);
}
```

File: dev/EnvironmentManager/API/Microsoft.Windows.System.EnvironmentManager.cpp
Comment (DrusTheAxe) [inline dev/EnvironmentManager/API/Microsoft.Windows.System.EnvironmentManager.cpp:29]: NIT: remove nullptr?

Does EnvironmentManager have a void ctor? If so then nullptr can be removed.
(if it doesn't but ctor(ptr=null) is legal then consider adding a void ctor :-)

File: test/EnvironmentManagerTests/EnvironmentVariableHelper.h
Comment (DrusTheAxe) [inline test/EnvironmentManagerTests/EnvironmentVariableHelper.h:168]: `UINT`? How very retro :P

`UINT32` or `uint32_t` is more hip these days

same comment applies throughout

File: test/EnvironmentManagerTests/EnvironmentVariableHelper.h
Comment (DrusTheAxe) [inline test/EnvironmentManagerTests/EnvironmentVariableHelper.h:216]: NIT: `...currentIndex{};`

No need for the explicit 0. The default initialization sets that

same comment applies throughout

**Review by DrusTheAxe — APPROVED** [review]
LGTM modulo comments noted

**Review by DrusTheAxe — APPROVED** [review]
LGTM modulo feedback

File: dev/EnvironmentManager/API/Microsoft.Windows.System.EnvironmentManager.Insights.h
Comment (dhoehna) [reply dev/EnvironmentManager/API/Microsoft.Windows.System.EnvironmentManager.Insights.h:9]: Now that I think about it, insights is redundant.  Should this go into Microsoft.WIndows.System.EnvironmentManager.h instead?

File: dev/EnvironmentManager/API/Microsoft.Windows.System.EnvironmentManager.Insights.h
Comment (ZenBird) [inline dev/EnvironmentManager/API/Microsoft.Windows.System.EnvironmentManager.Insights.h:41]: >TraceLoggingClassWrite [](http://example.com/codeflow?start=8&length=22)

This will not log the PartB information. Any reason why you are calling TraceLoggingClassWrite explicitly instead of defining an event (CustomEvent) and logging the data as payload?

File: dev/EnvironmentManager/API/Microsoft.Windows.System.EnvironmentManager.cpp
Comment (dhoehna) [reply dev/EnvironmentManager/API/Microsoft.Windows.System.EnvironmentManager.cpp:29]: Yep.  nullptr constructor is made by default.  In fact, this has to be done in this case.

https://docs.microsoft.com/en-us/windows/uwp/cpp-and-winrt-apis/author-apis#if-youre-authoring-a-runtime-class-in-a-windows-runtime-component

Whether you have IDL-declared constructor(s) or not, a constructor overload that takes std::nullptr_t is generated on your projected type. **Calling the std::nullptr_t constructor is the first of two steps in consuming the runtime class from the same compilation unit.** For more details, and a code example, see Consume APIs with C++/WinRT.

File: test/EnvironmentManagerTests/EnvironmentVariableHelper.h
Comment (dhoehna) [reply test/EnvironmentManagerTests/EnvironmentVariableHelper.h:168]: I don't want to be retro.

File: dev/EnvironmentManager/API/Microsoft.Windows.System.EnvironmentManager.cpp
Comment (dhoehna) [reply dev/EnvironmentManager/API/Microsoft.Windows.System.EnvironmentManager.cpp:29]: In fact.  If you try anything else the compiler will get stuck in a loop.

File: dev/EnvironmentManager/API/Microsoft.Windows.System.EnvironmentManager.Insights.h
Comment (DrusTheAxe) [reply dev/EnvironmentManager/API/Microsoft.Windows.System.EnvironmentManager.Insights.h:9]: Your call. Depends if you want your insights support in its own file or in the same as other stuff.

File: dev/EnvironmentManager/API/Microsoft.Windows.System.EnvironmentManager.Insights.h
Comment (dhoehna) [reply dev/EnvironmentManager/API/Microsoft.Windows.System.EnvironmentManager.Insights.h:41]: image

By partB i'm guessing you mean the message?  Both are logged.

I'm calling TraceLoggingClassWrite because that is how I saw how to call a custom event.  Is there a better way to call a custom event?

I could convert the scope to a string and add it to a std::wstring and call StringTelemetryEvent and that way there isn't a need for a customer event.

Is there a different way to call a custom event?

Comment (ZenBird) [reply dev/EnvironmentManager/API/Microsoft.Windows.System.EnvironmentManager.Insights.h:41]: Nope. PartB contains the information about the SDK and other fields that must be logged with all data coming from WindowsAppSDK. This will help us produce metrics on the usage and stability of code in different versions of WindowsAppSDK.

Comment (ZenBird) [reply dev/EnvironmentManager/API/Microsoft.Windows.System.EnvironmentManager.Insights.h:41]: You could have a member similar to m_scope like m_scopeString and log that as part of the payload in the event. Alternately, you could log the enum (as int) as part of the payload.

File: dev/EnvironmentManager/API/Microsoft.Windows.System.EnvironmentManager.Insights.h
Comment (ZenBird) [inline dev/EnvironmentManager/API/Microsoft.Windows.System.EnvironmentManager.Insights.h:17]: >ScopeTelemetryEvent [](http://example.com/codeflow?start=44&length=19)

Where is this event being logged? Isn't this event identical to the LogWithScopeAndMessage function with the difference that LogWithScopeAndMessage converts the scope to string while this logs the scope as an enum.

Comment (dhoehna) [reply dev/EnvironmentManager/API/Microsoft.Windows.System.EnvironmentManager.Insights.h]: You commented before I pushed. 

This was an attempt to log the scope and string before I moved onto the custom event.  It is gone now and replaced with a customer even "LogMessageAndScope"

**Review by ZenBird — APPROVED** [review]
:shipit:

**Pull request merged**
